core/src/main/java/io/grpc/internal/KeepAliveManager.java

```diff
@@ -170,8 +171,10 @@ public synchronized void onDataReceived() {
       }
       // schedule a new ping
       state = State.PING_SCHEDULED;
-      pingFuture =
-          scheduler.schedule(sendPing, keepAliveTimeInNanos, TimeUnit.NANOSECONDS);
+      if (pingFuture == null) {
```

Comment:
@stephenh Your argument makes sense. We should stay `IDLE` when we are in `IDLE` while ping had been scheduled, and then `onDataReceived`. But the check `if (pingFuture == null)` still looks strange, may be `Preconditions.checkState(pingFuture == null, "There should be no outstanding pingFuture")` at this place.